src/ripple/basics/SHA512Half.h

```diff
@@ -0,0 +1,123 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2014 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_BASICS_SHA512HALF_H_INCLUDED
+#define RIPPLE_BASICS_SHA512HALF_H_INCLUDED
+
+#include <ripple/basics/base_uint.h>
+#include <ripple/basics/Slice.h> // needed by most callers
+#include <beast/crypto/sha512.h>
+#include <beast/hash/hash_append.h>
+#include <beast/cxx14/type_traits.h> // <type_traits>
+
+namespace ripple {
+
+namespace detail {
+
+template <bool Secure>
+class SHA512HalfHasher
+{
+private:
+    using hasher_type =
+        std::conditional_t<Secure,
+            beast::sha512_hasher_s,
+                beast::sha512_hasher>;
+
+    hasher_type hasher_;
+
+public:
+    static beast::endian const endian =
+        beast::endian::big;
+
+    using result_type = uint256;
+
+    void
+    operator() (void const* data,
+        std::size_t size) noexcept
+    {
+        hasher_(data, size);
+    }
+
+    result_type
+    finish() noexcept
+    {
+        result_type digest;
+        auto const result = static_cast<
+            typename decltype(hasher_)::result_type>
+                (hasher_);
+        std::memcpy(digest.data(),
+            result.data(), 32);
```

Comment:
That's fine. My concern is only that if the type of `result_type` ever has cause to change, then it'll be easy to overlook the 32. And it's a very small concern, because I doubt it'll ever change.